ipa-core/src/protocol/ipa_prf/shuffle/mod.rs

```diff
@@ -0,0 +1,167 @@
+    },
+};
+
+pub mod base;
+
+#[tracing::instrument(name = "shuffle_inputs", skip_all)]
+async fn shuffle_inputs<C, BK, TV, TS>(
+    ctx: C,
+    input: Vec<OprfReport<BK, TV, TS>>,
+) -> Result<Vec<OprfReport<BK, TV, TS>>, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<Boolean>: UpgradedContext<Boolean, Share = AdditiveShare<Boolean>>,
+    BK: SharedValue + CustomArray<Element = Boolean> + Field,
+    TV: SharedValue + CustomArray<Element = Boolean> + Field,
+    TS: SharedValue + CustomArray<Element = Boolean> + Field,
+    for<'a> &'a AdditiveShare<TS>: IntoIterator<Item = AdditiveShare<Boolean>>,
+    for<'a> &'a AdditiveShare<TV>: IntoIterator<Item = AdditiveShare<Boolean>>,
+    for<'a> &'a AdditiveShare<BK>: IntoIterator<Item = AdditiveShare<Boolean>>,
+    for<'a> <&'a AdditiveShare<TV> as IntoIterator>::IntoIter: Send,
+    for<'a> <&'a AdditiveShare<TS> as IntoIterator>::IntoIter: Send,
+{
+    let shuffle_input: Vec<AdditiveShare<BA112>> = input
+        .into_iter()
+        .map(|item| oprfreport_to_shuffle_input::<BA112, BK, TV, TS>(&item))
+        .collect::<Vec<_>>();
+
+    let shuffled = shuffle(ctx, shuffle_input).await?;
+
+    Ok(shuffled
+        .into_iter()
+        .map(|item| shuffled_to_oprfreport(&item))
+        .collect::<Vec<_>>())
+}
+
+// This function converts OprfReport to an AdditiveShare needed for shuffle protocol
+pub fn oprfreport_to_shuffle_input<YS, BK, TV, TS>(
+    input: &OprfReport<BK, TV, TS>,
+) -> AdditiveShare<YS>
+where
+    YS: CustomArray<Element = <BA112 as CustomArray>::Element> + SharedValue,
+    BK: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    TV: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    TS: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+{
+    let (mut y_left, mut y_right) = (YS::ZERO, YS::ZERO);
+    expand_array_in_place(&mut y_left, input.match_key.left(), 0);
+    expand_array_in_place(&mut y_right, input.match_key.right(), 0);
+
+    let mut offset = BA64::BITS as usize;
+
+    y_left.set(offset, input.is_trigger.left());
+    y_right.set(offset, input.is_trigger.right());
+
+    offset += 1;
+
+    expand_array_in_place(&mut y_left, input.breakdown_key.left(), offset);
+    expand_array_in_place(&mut y_right, input.breakdown_key.right(), offset);
+
+    offset += BK::BITS as usize;
+    expand_array_in_place(&mut y_left, input.trigger_value.left(), offset);
+    expand_array_in_place(&mut y_right, input.trigger_value.right(), offset);
+
+    offset += TV::BITS as usize;
+    expand_array_in_place(&mut y_left, input.timestamp.left(), offset);
+    expand_array_in_place(&mut y_right, input.timestamp.right(), offset);
+
+    AdditiveShare::<YS>::new(y_left, y_right)
+}
+
+// This function converts AdditiveShare obtained from shuffle protocol to OprfReport
+pub fn shuffled_to_oprfreport<YS, BK, TV, TS>(input: &AdditiveShare<YS>) -> OprfReport<BK, TV, TS>
```

Comment:
I have noticed that we have a lot of structs that are more or less representing the same data. I was wondering whether we can get rid of some of the structs like OPRF report. I was thinking that the decryption just outputs a bunch of u128s (or [u8, 16] from AES) we can convert that in a large BA containing mk, tb, bk, tv, ts. We could directly shuffle the BA without converting them to a oprfreport. Then for evaluating the prf, you just call it on a stream over the BAs that extracts the mk from the BA and overwrites the mk with its pseudonym (we could make sure mk and pseudonym use the same BA type. For the sort according to pseudonym you also use the BA type as input. After all of that you could finally drop the pseudonym and extract the rest in the BA, i.e. tb, bk, tv, ts, and pass it into the attribution circuit. This way we wouldn't need all the conversions between the different struct types which might cost reallocations (maybe the compiler is able to optimize them away?).

Reply:
I agree. This seems like a potential optimization that may come at the cost of readbility. Let's revisit that later in the optimization stage.